src/cached_object_store.rs

```diff
@@ -503,12 +512,15 @@ pub(crate) enum InvalidGetRange {
     Inconsistent { start: usize, end: usize },
 }
 
+#[async_trait::async_trait]
 pub trait LocalCacheStorage: Send + Sync + std::fmt::Debug + Display + 'static {
     fn entry(
         &self,
         location: &object_store::path::Path,
         part_size: usize,
     ) -> Box<dyn LocalCacheEntry>;
+
+    async fn start_evictor(&self);
```

Comment:
good idea. going with an "independent" evictor not coupled with the cache implementation is the way to go; this modularizes the cache component of the system. if/when there are concrete cases where tightly coupling an evictor to a specific cache implementation makes the most sense, it can be dealt with as a special case at that point.

Reply:
I am not very sure whether it will introduce additional complexities to the `LocalCacheStorage` trait if we want to add some support methods to the evictors. in the earlier stage, I think it might be not bad to keep the API in the `LocalCacheStorage` trait as small as possible until we learn more about the possible extension needs.

however, I think it's nice to have a try at extracting the implementation of 2-choice as a generic algorithm here. so when we have other cache storages impls, they could choose to call this helper algorithm, instead of exposing more APIs or copy paste the implementation code.

Reply:
after some thought, i am in doubt about whether making it generic at this early stage is worthwhile.

in the mid term, the LocalCacheStorage impls are expected to be limitted (fs and rocksdb), and the eviction algorithms both actually couples with the underlying storage in these two cases:

**Filesystem cache storage**: this is expected to serve as a good default setup for most use cases. However, it might suffer when dealing with a large number of small files, especially if the cache size is substantial. In the filesystem, the eviction algorithms are primarily an approximation of LRU (Least Recently Used) with randomization.

**RocksDB cache storage**: if you want to avoid battling with huge amount of small files, you can opt for RocksDB's BlobDB with a FIFO strategy. FIFO is very trivial to implement with BlobDB, and the evictor simply does nothing but a rocksdb config. while it's possible to implement a choice-of-two algorithm with RocksDB, this approach might negate the benefits of no compaction and no write amplification offered by BlobDB with FIFO.
 
as described, implementing FIFO with filesystem storage is challenging, while implementing a choice-of-two algorithm with RocksDB storage may not be worthwhile. it might becomes hard to make a generic eviction algorithm for these both :(

I guess it might still be possible to extract common code when we encounter code duplication in the future. what are your thoughts on this?

Reply:
SGTM! Let's not over complicate it. :)